api/firmware/btc.go

```diff
@@ -240,20 +243,64 @@ func (device *Device) BTCSign(
 	if err != nil {
 		return nil, err
 	}
+
+	isInputsPass2 := false
 	for {
 		switch next.Type {
 		case messages.BTCSignNextResponse_INPUT:
 			inputIndex := next.Index
+			input := *tx.Inputs[inputIndex].Input
+
+			var hostNonce []byte
+			if supportsAntiklepto && isInputsPass2 {
+				hostNonce = bytesOrPanic(32)
+				input.HostNonceCommitment = &messages.AntiKleptoHostNonceCommitment{
+					Commitment: antikleptoHostCommit(hostNonce),
+				}
+			}
 			next, err = device.queryBtcSign(&messages.Request{
 				Request: &messages.Request_BtcSignInput{
-					BtcSignInput: tx.Inputs[inputIndex].Input,
+					BtcSignInput: &input,
 				}})
 			if err != nil {
 				return nil, err
 			}
-			if next.HasSignature {
+
+			if supportsAntiklepto && isInputsPass2 {
+				if next.Type != messages.BTCSignNextResponse_HOST_NONCE || next.AntiKleptoSignerCommitment == nil {
```

Comment:
Why not add this as the switch case to `switch next.Type { .. }`. I think it would be easier to read and follow the protocol:

```go

var antikleptoErr = errors.New("unverified antiklepto")
for {
  switch next.Type {
  case messages.BTCSignNextResponse_INPUT:
  ...
  case messages.BTCSignNextResponse_HOST_NONCE:
   antikleptoErr = antikleptoVerify(...)
  }
  ...prevtx
}

if supportsAntiklepto && antikleptoErr != nil {
  return antikleptoErr
}
```

Reply:
Discussed in person  - I want to switch this to be explicit (not next-based) anyway, and also we have to be sure of the right order here (send host nonce only after receiving the client commitment), which would be a bit harder.